src/WebJobs.Script/Description/FunctionInvokerBase.cs

```diff
@@ -81,7 +86,80 @@ protected bool InitializeFileWatcherIfEnabled()
             return false;
         }
 
-        public abstract Task Invoke(object[] parameters);
+        public async Task Invoke(object[] parameters)
+        {
+            FunctionStartedEvent startedEvent = null;
+
+            // We require the ExecutionContext, so this will throw if one is not found.
+            ExecutionContext functionExecutionContext = (ExecutionContext)parameters.First(p => p?.GetType() == typeof(ExecutionContext));
+
+            // These may not be present, so null is okay.
+            TraceWriter functionTraceWriter = (TraceWriter)parameters.FirstOrDefault(p => p is TraceWriter);
+            Binder binder = (Binder)parameters.FirstOrDefault(p => p?.GetType() == typeof(Binder));
+
+            string invocationId = functionExecutionContext.InvocationId.ToString();
+
+            startedEvent = new FunctionStartedEvent(functionExecutionContext.InvocationId, Metadata);
+            _metrics.BeginEvent(startedEvent);
+
+            try
+            {
+                TraceWriter.Info($"Function started (Id={invocationId})");
+
+                FunctionInvocationContext context = new FunctionInvocationContext
+                {
+                    ExecutionContext = functionExecutionContext,
+                    Binder = binder,
+                    TraceWriter = functionTraceWriter
```

Comment:
What's the difference between context.TraceWriter and the TraceWriter property used below?